### PR TITLE
EDSC-2450: Implemented Datepicker and DatepickerContainer with the expected functionality

### PR DESCRIPTION
**Pull request by bnp26**
# Overview

### What is the feature?

Changed the Datepicker/DatepickerContainer so that when closing the calendar/input (blurring), dates autocomplete depending on the datetime format that was set in the input.

### What is the Solution?

Updated Datepicker, DatepickerContainer, and their tests (all associated tests that are impacted by the change)

### What areas of the application does this impact?

List impacted areas.

# Testing

### Reproduction steps

- **Environment for testing:**
- **Collection to test with:**

1. Go to localhost:8080/search
2. Open the temporal datetime filter.
3. Enter 1999 in the start & end inputs
4. observe the fields autocompleting to 1999-01-01 00:00:00 & 1999-12-31 23:59:59 respectively
5. Enter 1999-02 in start and end inputs
6. observe the fields autocompleting to 1999-02-01 00:00:00 & 1999-02-28 23:59:59 respectively
7. edit the start date to 1999-01-06 12:00:00 and end date to 1999-02-06 13:00:00
8. observe the dates don't autocomplete
9. select a date using the calendar for the start and end dates and observe they autocomplete correctly.
### Attachments

Please include relevant screenshots or files that would be helpful in reviewing and verifying this change.

We have good coverage on tests when run locally.
Screenshot 2024-08-13 at 1 16 11 PM
Screenshot 2024-08-13 at 1 16 34 PM

# Checklist

- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

**Review by daniel-zamora — APPROVED** [review]

File: static/src/js/containers/DatepickerContainer/DatepickerContainer.jsx [hunk 1]
```diff
@@ -34,20 +34,29 @@ class DatepickerContainer extends Component {
...
-  onBlur() {
-    const { viewMode } = this.props
+  * Autocomplete the field on blur
+  */
+  onInputBlur(e) {
+    const {
+      format
+    } = this.props
+
+    console.log(e)
```
Comment (eudoroolivares2016) [inline static/src/js/containers/DatepickerContainer/DatepickerContainer.jsx:44]: big log in the console

File: static/src/js/components/GranuleFilters/GranuleFiltersForm.jsx [hunk 2]
```diff
@@ -445,7 +445,7 @@ export const GranuleFiltersForm = (props) => {
                       value: isChecked
                     })
 
-                    // This timeout is needed for there to be no delay in the checkbox working
+                    // Previously there was a timeout here to await the event loop
```
Comment (trevorlang) [inline static/src/js/components/GranuleFilters/GranuleFiltersForm.jsx:448]: Do we need this comment anymore?

File: static/src/js/components/TemporalDisplay/TemporalSelectionDropdown.jsx [hunk 3]
```diff
@@ -181,6 +181,15 @@ const TemporalSelectionDropdown = ({
       newStartDate.year(startDateObject.year())
     }
 
+    if (!newStartDate) {
```
Comment (trevorlang) [inline static/src/js/components/TemporalDisplay/TemporalSelectionDropdown.jsx:184]: Is this change still necessary?

File: static/src/js/components/TemporalDisplay/TemporalSelectionDropdown.jsx [hunk 4]
```diff
@@ -198,6 +207,15 @@ const TemporalSelectionDropdown = ({
       isRecurring: existingIsRecurring
     } = temporal
 
+    if (!newEndDate) {
```
Comment (trevorlang) [inline static/src/js/components/TemporalDisplay/TemporalSelectionDropdown.jsx:210]: Is this change still necessary?

File: static/src/js/components/TemporalDisplay/TemporalSelectionDropdownToggle.jsx [hunk 5]
```diff
@@ -9,6 +9,7 @@ const TemporalSelectionDropdownToggle = ({ onToggleClick }) => (
   <Dropdown.Toggle
     variant="inline-block"
     id="temporal-selection-dropdown"
+    aria-label="temporal-selection-dropdown"
```
Comment (trevorlang) [inline static/src/js/components/TemporalDisplay/TemporalSelectionDropdownToggle.jsx:12]: This is not a good aria-label. This should be some human readable text. It is meant to add context to users with assistive technology. Something like "Open temporal filters" might work well here.

File: static/src/js/containers/DatepickerContainer/__tests__/DatepickerContainer.test.js [hunk 6]
```diff
@@ -79,41 +175,52 @@ describe('DatepickerContainer component', () => {
...
+      // This is because the date will become endOf('day') on Blur
+      expect(onChange).toHaveBeenCalledWith('1988-09-03 23:59:59')
 
       MockDate.reset()
     })
   })
 
   describe('onClearClick', () => {
-    test('triggers the onChange function with an empty string', () => {
+    test('does not trigger the onChange and calls onSubmit with a null value', () => {
```
Comment (trevorlang) [inline static/src/js/containers/DatepickerContainer/__tests__/DatepickerContainer.test.js:212]: I dont think this test name makes sense

File: static/src/js/containers/DatepickerContainer/__tests__/DatepickerContainer.test.js [hunk 7]
```diff
@@ -167,11 +274,15 @@ describe('DatepickerContainer component', () => {
...
-        const { onChange } = enzymeWrapper.instance()
+
+        enzymeWrapper.instance().picker.current = {}
+        enzymeWrapper.instance().picker.current.setState = jest.fn()
+
+        const { onInputBlur } = enzymeWrapper.instance()
         const { onSubmit } = props
         enzymeWrapper.setProps({ type: 'start' })
 
-        onChange(moment.utc('1988-09-03'))
```
Comment (trevorlang) [inline static/src/js/containers/DatepickerContainer/__tests__/DatepickerContainer.test.js:174]: I dont see why this wouldve needed to be swapped to onInputBlur. Arent we still ultimately calling onChange with the values?

File: static/src/js/containers/DatepickerContainer/__tests__/DatepickerContainer.test.js [hunk 8]
```diff
@@ -183,11 +294,15 @@ describe('DatepickerContainer component', () => {
...
+
+        enzymeWrapper.instance().picker.current = {}
+        enzymeWrapper.instance().picker.current.setState = jest.fn()
+
+        const { onInputBlur } = enzymeWrapper.instance()
         const { onSubmit } = props
         enzymeWrapper.setProps({ type: 'end' })
 
-        onChange(moment.utc('1988-09-03'))
+        onInputBlur({ target: { value: moment.utc('1988-09-03') } })
```
Comment (trevorlang) [inline static/src/js/containers/DatepickerContainer/__tests__/DatepickerContainer.test.js:305]: I dont see why this wouldve needed to be swapped to onInputBlur. Arent we still ultimately calling onChange with the values?

File: static/src/js/containers/GranuleFiltersContainer/GranuleFiltersContainer.jsx [hunk 9]
```diff
@@ -92,7 +92,11 @@ export const GranuleFiltersContainer = (props) => {
 
   const onHandleSubmit = () => {
     // Pass the internal `Formik` values
-    handleSubmit(values)
+    // Waits for the next event cycle to execute
+    // Allows validation for temporal selection to work as expected
+    setTimeout(() => {
```
Comment (trevorlang) [inline static/src/js/containers/GranuleFiltersContainer/GranuleFiltersContainer.jsx:97]: This change scares me. @eudoroolivares2016 you added this correct? Is this to bypass a side-effect of the EDSC-4218 bug?

File: static/src/js/components/GranuleFilters/GranuleFiltersForm.jsx (hunk 2 above)
Comment (eudoroolivares2016) [reply static/src/js/components/GranuleFilters/GranuleFiltersForm.jsx:448]: Ehh I would rather keep this because, I cannot explain why we used to have a setTimeout there prior to the granule filter update I made. If you really want I will remove it

File: static/src/js/containers/GranuleFiltersContainer/GranuleFiltersContainer.jsx (hunk 9 above)
Comment (eudoroolivares2016) [reply static/src/js/containers/GranuleFiltersContainer/GranuleFiltersContainer.jsx:97]: I added this because this [commit](https://github.com/nasa/earthdata-search/commit/0944b7229d46144dd74cda601509337d7d695d4e) introduced an issue where the validation was happening after submission.

This is not a new change. That settimeout was originally on the function for the last 5 years prior to that commit above. I had thought the issue would be fixed by this change and was caused by Vite when I made it but, its coming from this.

File: static/src/js/components/GranuleFilters/GranuleFiltersForm.jsx (hunk 2 above)
Comment (trevorlang) [reply static/src/js/components/GranuleFilters/GranuleFiltersForm.jsx:448]: Its just odd to comment code that isn't there anymore. We should not be making changes we don't feel comfortable with. Leaving something like this behind feels like a red flag.

**Comment by trevorlang** [comment]
@bnp26 This branch needs to be updated with `main`

File: static/src/js/components/TemporalDisplay/TemporalSelectionDropdown.jsx
Comment (trevorlang) [inline static/src/js/components/TemporalDisplay/TemporalSelectionDropdown.jsx:1]: Looks like you removed the `setTemporal` change from `setStartDate` but not `setEndDate`. Was that on purpose? It seemed to me like it could be removed from both.

File: static/src/js/containers/DatepickerContainer/DatepickerContainer.jsx [hunk 10]
```diff
@@ -43,6 +43,8 @@ class DatepickerContainer extends Component {
 
     const { value } = e.target
 
+    console.log(value)
```
Comment (trevorlang) [inline static/src/js/containers/DatepickerContainer/DatepickerContainer.jsx:46]: Remove

File: static/src/js/containers/DatepickerContainer/DatepickerContainer.jsx [hunk 11]
```diff
@@ -71,6 +73,7 @@ class DatepickerContainer extends Component {
     } = this.props
 
     let valueToSet = null
+    console.log(value)
```
Comment (trevorlang) [inline static/src/js/containers/DatepickerContainer/DatepickerContainer.jsx:76]: Remove

File: static/src/js/containers/DatepickerContainer/DatepickerContainer.jsx [hunk 12]
```diff
@@ -82,7 +85,16 @@ class DatepickerContainer extends Component {
       if (type === 'start') {
         valueToSet = value.startOf('day')
       } else if (type === 'end') {
-        valueToSet = value.endOf('day')
+        // Using moment.ISO_8601 to determine format of the user input
+        const iso2806Date = moment.utc(value.creationData().input, moment.ISO_8601)
+        const { format: dateFormat } = iso2806Date.creationData()
+        if (dateFormat === 'YYYY') {
```
Comment (trevorlang) [inline static/src/js/containers/DatepickerContainer/DatepickerContainer.jsx:91]: Add a return between definitions and `if` blocks

File: static/src/js/containers/DatepickerContainer/__tests__/DatepickerContainer.test.js [hunk 13]
```diff
@@ -82,15 +82,50 @@ describe('DatepickerContainer component', () => {
 
       // Because this is a shallow render the Datepicker component doesn't render so we have
       // to manually create a 'current' instance of the component
-      enzymeWrapper.instance().picker.current = {}
-      enzymeWrapper.instance().picker.current.setState = jest.fn()
+      // enzymeWrapper.instance().picker.current = {}
```
Comment (trevorlang) [inline static/src/js/containers/DatepickerContainer/__tests__/DatepickerContainer.test.js:85]: Remove commented code

File: static/src/js/containers/DatepickerContainer/__tests__/DatepickerContainer.test.js [hunk 14]
```diff
@@ -82,15 +82,50 @@ describe('DatepickerContainer component', () => {
...
+      const { onSubmit, format } = props
+
+      const onChangeSpy = jest.spyOn(enzymeWrapper.instance(), 'onChange')
+
+      enzymeWrapper.instance().onInputBlur({ target: { value: '1990' } })
+
+      expect(onChangeSpy).toHaveBeenCalledTimes(1)
+      expect(onSubmit).toHaveBeenCalledTimes(1)
+      // Had trouble matching the actual input that was being passed to the function
+      expect(onSubmit.mock.calls[0][0].format(format)).toEqual(moment.utc('1990', format).endOf('year').format(format))
```
Comment (trevorlang) [inline static/src/js/containers/DatepickerContainer/__tests__/DatepickerContainer.test.js:105]: `moment.utc('1990', format).endOf('year').format(format)` just returns a string so we should check against that resulting string value in this `toEqual` to make things a little more simple.

File: static/src/js/containers/DatepickerContainer/__tests__/DatepickerContainer.test.js [hunk 15]
```diff
@@ -82,15 +82,50 @@ describe('DatepickerContainer component', () => {
...
+      enzymeWrapper.instance().picker.current.setState = jest.fn()
+
+      const onChangeSpy = jest.spyOn(enzymeWrapper.instance(), 'onChange')
+
+      onInputBlur({ target: { value: '1990-06' } })
+
+      expect(onChangeSpy).toHaveBeenCalledTimes(1)
+      expect(onSubmit).toHaveBeenCalledTimes(1)
+      // Had trouble matching the actual input that was being passed to the function
+      expect(onSubmit.mock.calls[0][0].format(format)).toEqual(moment.utc('1990-06', format).endOf('month').format(format))
```
Comment (trevorlang) [inline static/src/js/containers/DatepickerContainer/__tests__/DatepickerContainer.test.js:126]: `moment.utc('1990-06', format).endOf('month').format(format)` just returns a string so we should check against that resulting string value in this toEqual to make things a little more simple.

**Review by trevorlang — CHANGES_REQUESTED** [review]

File: static/src/js/components/TemporalDisplay/TemporalSelectionDropdown.jsx
Comment (bnp26) [reply static/src/js/components/TemporalDisplay/TemporalSelectionDropdown.jsx:1]: No, just forgot

**Review by trevorlang — APPROVED** [review]
